ui/src/app/pages/homepage/data-access-requirements.tsx

```diff
@@ -0,0 +1,531 @@
+      [!eligible, () => <MinusCircle data-test-id={`module-${module}-ineligible`} style={{color: colors.disabled}}/>],
+      // eligible and (completed or bypassed)
+      [eligible && !!statusTextMaybe, () => <CheckCircle data-test-id={`module-${module}-complete`} style={{color: colors.success}}/>],
+      // eligible and incomplete and unbypassed
+      [eligible && !statusTextMaybe, () => <CheckCircle data-test-id={`module-${module}-incomplete`} style={{color: colors.disabled}}/>])}
+  </div>;
+
+  const ModuleBox = ({children}) => {
+    return active ?
+        <Clickable onClick={() => {
+          setShowRefresh(true);
+          moduleAction();
+        }}>
+          <FlexRow style={styles.activeModuleBox}>{children}</FlexRow>
+        </Clickable> :
+        <FlexRow style={styles.inactiveModuleBox}>{children}</FlexRow>;
+  };
+
+  const Module = () => <FlexRow data-test-id={`module-${module}`}>
+    <FlexRow style={styles.moduleCTA}>
+      {active && (showRefresh ? <Refresh/> : <Next/>)}
+    </FlexRow>
+    <ModuleBox>
+      <ModuleIcon/>
+      <FlexColumn>
+        <div style={active ? styles.activeModuleText : styles.inactiveModuleText}>
+          {moduleLabels.get(module)}
+        </div>
+        {statusTextMaybe && <div style={styles.moduleDate}>{statusTextMaybe}</div>}
+      </FlexColumn>
+    </ModuleBox>
+    {showTwoFactorAuthModal && <TwoFactorAuthModal
+        onClick={() => setShowTwoFactorAuthModal(false)}
+        onCancel={() => setShowTwoFactorAuthModal(false)}/>}
+  </FlexRow>;
+
+  const moduleEnabled = !!registrationTask;
+  return moduleEnabled ? <Module/> : null;
+};
+
+export const DataAccessRequirements = fp.flow(withProfileErrorModal)((spinnerProps: WithSpinnerOverlayProps) => {
+  const {profile, reload} = useStore(profileStore);
+
+  const syncExternalModulesAndReloadProfile = async() => {
+    const aborter = new AbortController();
+    spinnerProps.showSpinner();
+    await Promise.all([
+      profileApi().syncTwoFactorAuthStatus({signal: aborter.signal}),
+      profileApi().syncComplianceTrainingStatus({signal: aborter.signal}),
+      profileApi().syncEraCommonsStatus({signal: aborter.signal}),
+    ]);
+    spinnerProps.hideSpinner();
+    reload();
+
+    // cleanup on unmount
+    return aborter.abort;
+  };
+
+  useEffect(() => {
+    syncExternalModulesAndReloadProfile();
+  }, []);
+
+  // handle the route /nih-callback?token=<token>
+  // handle the route /ras-callback?code=<code>
+  const {token, code} = queryParamsStore.getValue();
+  useEffect(() => {
+    if (token) {
+      handleTerraShibbolethCallback(token, spinnerProps);
+    }
+  }, [token] );
+  useEffect(() => {
+    if (code) {
+      handleRasCallback(code, spinnerProps);
+    }
+  }, [code] );
+
+  // which module are we currently guiding the user to complete?
+  const [activeModule, setActiveModule] = useState(null);
+
+  const [navigate, ] = useNavigation();
+  const enabledModules = allModules.map(module => {
```

Comment:
nit: sort of awkward that you'd have nulls in this array. Might be cleaner to do a `flatMap` to map/filter in one go, something like:

```
if (!enabledTaskMaybe) {
  return [];
}
return [enabledTaskMaybe.module];
```

Reply:
Oops.  That was my intention but I made a false assumption about how much magic was getting applied here.  Thanks.

Aside: I had to fire up a local API to test this, and this process was ~4x faster thanks to your recent docker fixes.